### PR TITLE
[feature] Allow Saxon licensed editions (EE and PE)

### PR DESCRIPTION
**Pull request by alanpaxton**
### Description:

* Add support for loading a Saxon configuration file from a location defined in eXist-db's `conf.xml`.
* Refactor current code in eXist-db which uses Saxon, to use Saxon configuration and Saxon processor provided by the broker pool (single configuration and processor per broker pool)
* Introduces a `SaxonConfigurationHolder` class used by the broker pool to search for and fetch the Saxon configuration referenced by `conf.xml` and load it into Saxon. The holder holds the saxon configuration lazily, loads it on first request, and retains the same configuration to respond to subsequent requests with.
* Provide instructions for replacing the SaxonHE bundled in eXist with a licensed Saxon (EE or PE). The changes to configuration loading allow the user to configure their use of a licensed version of Saxon so that xquery functions such as `fn:transform()` (which uses Saxon as its transformer) can make use of EE and/or PE features.

### Reference:

This feature was requested in https://github.com/eXist-db/exist/issues/4813

### Type of tests:

* Basic unit tests to verify the loading of Saxon configuration via the broker pool.
* The manual procedure to use EE has been validated by installing a clean eXist from an installer Jar built from the feature branch, walking through the instructions and executing a `fn:transform()` script in eXide. Logging by the configuration holder of the result of querying capabilities of the resulting Saxon processor verified that the running version of Saxon provided all licensed features.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public final class SaxonConfigurationHolder {
+
+  public final static String CONFIGURATION_ELEMENT_NAME = "saxon";
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java:42]: Generic name of static export should be avoided.
see also comment from @adamretter in #4838

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  public final static String CONFIGURATION_FILE_ATTRIBUTE = "configuration-file";
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java:43]: Too generic name, see above

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+  public final static String CONFIGURATION_FILE_PROPERTY = "saxon.configuration";
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java:44]: Too generic name, see above

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    this.brokerPool = brokerPool;
+  }
+
+  /**
+   * Factory for the singleton Saxon configuration wrapper per broker pool
+   * @param brokerPool for which to fetch (and if necessary create) a Saxon configuration
+   *
+   * @return the associated Saxon configuration wrapper
+   */
+  public synchronized static SaxonConfigurationHolder GetHolderForBroker(final BrokerPool brokerPool) {
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java:80]: Why did you choose to start the static method name with an uppercase letter?

File: exist-core/src/main/java/org/exist/util/Configuration.java [hunk 5]
```diff
@@ -260,6 +260,12 @@ public Configuration(String configFilename, Optional<Path> existHomeDirname) thr
                 configureTransformer((Element)transformers.item(0));
             }
 
+            //saxon settings (most importantly license file for PE or EE features)
+            final NodeList saxon = doc.getElementsByTagName(SaxonConfigurationHolder.CONFIGURATION_ELEMENT_NAME);
```
Comment (line-o) [inline exist-core/src/main/java/org/exist/util/Configuration.java:264]: Here you can see clearly that the name is too generic and clashes with other `CONFIGURATION_ELEMENT_NAME`s

**Review by line-o — CHANGES_REQUESTED** [review]
Nice Feature! Please have a look at the Codacy output and the exported names.
I also do wonder how this change affects the runtime of `fn:transform`.
I hope the constructor of this function does only run once and then reuses the static context and configuration within on each call. Is that the case?

**Comment by line-o** [comment]
@alanpaxton The commit messages do not conform to our commit message format. They must start with either `[enhancement]`, `[bugfix]`, `[refactor]` or `[ignore]`.

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java (hunk 4 above)
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java:80]: My brain seems to believe that is the standard. I can change it.

**Review by line-o — APPROVED** [review]
Thanks for going the extra-mile @alanpaxton !

**Comment by line-o** [comment]
@alanpaxton Can you rename your commits to match the commit message format?

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java (hunk 1 above)
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java]: I've prefixed all the statics with SAXON_

File: exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java (hunk 4 above)
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/util/SaxonConfigurationHolder.java]: Changed to camel.

**Review by line-o — APPROVED** [review]

**Comment by line-o** [comment]
Having learned that it was possible to use Saxon PE and EE up until 6.0.1 I see this as a fix of a regression. I believe this warrants a backport to develop-6.x.x

**Comment by adamretter** [comment]
@line-o we won't be backporting this ourselves, but anyone else is free to do so

**Review by adamretter — CHANGES_REQUESTED** [review]
Just some minor issues, and one question about the design of the config holder class.

**Comment by kosek** [comment]
Because of a bug in Saxon 9.9 that has been fixed only for version 10+ it is not possible to integrate Saxon-PE, only Saxon-EE.

See corresponding Saxon bug: https://saxonica.plan.io/issues/4849

I'm not sure whether this bug will be ever fixed in 9.9. Are there any plans/estimates to use Saxon 10/11 in exist-db?

File: exist-distribution/Saxon-EE.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# How to use Saxon PE or EE
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-distribution/Saxon-EE.md]: Removed from here now. Change in progress on the doc repo.

File: exist-core/src/main/java/org/exist/storage/BrokerPool.java [hunk 7]
```diff
@@ -379,7 +378,7 @@ public String getStatus() {
      *
      * One instance per-database, lazily initialised.
      */
-    private AtomicLazyVal<net.sf.saxon.Configuration> saxonConfig = new AtomicLazyVal<>(net.sf.saxon.Configuration::newConfiguration);
+    private SaxonConfigurationHolder saxonConfigurationHolder;
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-core/src/main/java/org/exist/storage/BrokerPool.java]: Refactored according to your suggestion.

File: exist-distribution/src/main/config/conf.xml [hunk 8]
```diff
@@ -1104,4 +1104,10 @@
     -->
     <xupdate allowed-fragmentation="50000" enable-consistency-checks="no"/>
 
+    <!-- 
+        Load this configuration into Saxon
+    -->
+    <saxon configuration-file="saxon-config.xml"/>
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-distribution/src/main/config/conf.xml:1114]: done

File: exist-distribution/src/main/config/saxon-config.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?><!-- 
+     This configuration file is placed here by eXist-DB as a placeholder.
+     It should be replaced if the user wishes to use a licensed edition of Saxon.
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-distribution/src/main/config/saxon-config.xml:3]: done

File: exist-core/src/test/resources-filtered/conf.xml [hunk 10]
```diff
@@ -931,4 +931,9 @@
     -->
     <xupdate allowed-fragmentation="50000" enable-consistency-checks="no"/>
 
+    <!-- 
+        Load this configuration ito Saxon
+    -->
```
[the remark this answers is not in the document]
Comment (alanpaxton) [reply exist-core/src/test/resources-filtered/conf.xml:936]: done

**Comment by alanpaxton** [comment]
> Because of a bug in Saxon 9.9 that has been fixed only for version 10+ it is not possible to integrate Saxon-PE, only Saxon-EE.
> 
> See corresponding Saxon bug: https://saxonica.plan.io/issues/4849
> 
> I'm not sure whether this bug will be ever fixed in 9.9. Are there any plans/estimates to use Saxon 10/11 in exist-db?

I had a look at upgrading to 11 when I started this, but saxon has changed the regex code which we fork as a separate repo and include in exist, so there's some quantity more work involved to update and resolve that conflict.

**Review by adamretter — APPROVED** [review]
Thanks for addressing the review comments - LGTM :-)

**Comment by adamretter** [comment]
@line-o Can this be merged now then?

**Pull request merged**